hail/src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -62,184 +65,322 @@ case class IndexMetadata(
-        makeLeafEnc,
-        makeIntEnc,
-        branchingFactor,
-        attributes)
+    val f = StagedIndexWriter.build(ctx, keyType, annotationType, branchingFactor, attributes);
+    { path: String =>
+      new IndexWriter(keyType, annotationType, f(path))
     }
   }
 }
 
-
-class IndexWriter(
-  fs: FS,
-  path: String,
-  keyType: PType,
-  annotationType: PType,
-  makeLeafEncoder: (OutputStream) => Encoder,
-  makeInternalEncoder: (OutputStream) => Encoder,
-  branchingFactor: Int = 4096,
-  attributes: Map[String, Any] = Map.empty[String, Any]) extends AutoCloseable {
-  require(branchingFactor > 1)
-
-  private var elementIdx = 0L
+class IndexWriter(keyType: PType, valueType: PType, comp: CompiledIndexWriter) extends AutoCloseable {
   private val region = Region()
   private val rvb = new RegionValueBuilder(region)
+  def +=(x: Annotation, offset: Long, annotation: Annotation): Unit = {
+    rvb.start(keyType)
+    rvb.addAnnotation(keyType.virtualType, x)
+    val koff = rvb.end()
+    rvb.start(valueType)
+    rvb.addAnnotation(valueType.virtualType, annotation)
+    val voff = rvb.end()
+    comp.apply(koff, offset, voff)
+  }
 
-  private val leafNodeBuilder = new LeafNodeBuilder(keyType, annotationType, 0L)
-  private val internalNodeBuilders = new ArrayBuilder[InternalNodeBuilder]()
-  internalNodeBuilders += new InternalNodeBuilder(keyType, annotationType)
-
-  private val trackedOS = new ByteTrackingOutputStream(fs.create(path + "/index"))
+  def close(): Unit = {
+    region.close()
+    comp.close()
+  }
+}
 
-  private val leafEncoder = makeLeafEncoder(trackedOS)
-  private val internalEncoder = makeInternalEncoder(trackedOS)
+class IndexWriterArrayBuilder(name: String, maxSize: Int, sb: SettableBuilder, region: Value[Region], arrayType: PCanonicalArray) {
+  private val aoff = sb.newSettable[Long](s"${name}_aoff")
+  private val len = sb.newSettable[Int](s"${name}_len")
 
-  private def height: Int = internalNodeBuilders.length + 1 // have one leaf node layer
+  val eltType: PCanonicalStruct = types.coerce[PCanonicalStruct](arrayType.elementType)
+  private val elt = new PCanonicalBaseStructSettable(eltType, sb.newSettable[Long](s"${name}_elt_off"))
 
-  private def writeInternalNode(node: InternalNodeBuilder, level: Int, isRoot: Boolean = false): Long = {
-    val indexFileOffset = trackedOS.bytesWritten
+  def length: Code[Int] = len
 
-    val info = if (node.size > 0) {
-      val firstChild = node.getChild(0)
-      val firstIndex = firstChild.firstIndex
-      val firstKey = firstChild.firstKey
-      val firstRecordOffset = firstChild.firstRecordOffset
-      val firstAnnotation = firstChild.firstAnnotation
-      IndexNodeInfo(indexFileOffset, firstIndex, firstKey, firstRecordOffset, firstAnnotation)
-    } else {
-      assert(isRoot && level == 0)
-      null
-    }
+  def loadFrom(cb: EmitCodeBuilder, a: Code[Long], l: Code[Int]): Unit = {
+    cb += (aoff := a)
+    cb += (len := l)
```

Comment:
prefer `assign` to `+= v := c`. We expect the latter form to go away, so changing it now is probably for the best.